Add ViaVersion packets support

Rubenicos opened the pull request:
> [!TIP]
> Known bugs:
> * ~~Somethings TAB send tablist and scoreboard information too early (before ViaVersion do all the stuff to properly send packets to players) causing to not see any of both, can be fixed for scoreboard if a delay is set, and for tablist only with `/tab reload`.~~

> [!NOTE]  
> Only Bukkit support by default, **I think** this can be used as well on Sponge module.

### Additions
* [Tablist] Hat update for +1.21.4 players on pre 1.21.4 servers.
* [Tablist] Text shadow color support for +1.21.4 players on pre 1.21.4 servers.
* [Tablist] List order support for +1.21.2 players on pre 1.21.2 servers.
* [Tablist] Listed update for +1.19.3 players on pre 1.19.3 servers.
* [Tablist] RGB colors support for +1.16 players on pre 1.16 servers.
* [Tablist] Any other stuff that I didn't mention above for +1.16 players on pre 1.16 servers.
* [Scoreboard] Text shadow color support for +1.21.4 players on pre 1.21.4 servers.
* [Scoreboard] Number format support for +1.20.3 players on pre 1.20.3 servers.
* [Scoreboard] RGB colors support for +1.16 players on pre 1.16 servers.
* [Scoreboard] Any other stuff that I didn't mention above for +1.13 players on pre 1.13 servers.

~~I used a separated module, as "viaversion platform" since it doesn't require more than a player id to send packets.~~

NEZNAMY commented:
Wouldn't it be better to have this in shared? There already are some existing implementations using adventure api for example.

NEZNAMY commented on component/src/main/java/me/neznamy/chat/TextColor.java:
```diff
@@ -125,6 +125,15 @@ private EnumChatFormat loadClosestColor() {
+     * @return true if this a legacy color, false otherwise.
+     */
+    public boolean isLegacy() {
+        return legacyColor != null;
```
legacyColor not being null can also mean the closest color was calculated on request.

NEZNAMY commented on viaversion/src/main/java/me/neznamy/tab/platforms/viaversion/scoreboard/ViaScoreboard.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Override
+    public @NotNull Object createTeam(@NonNull String name) {
+        return null;
```
NotNull

NEZNAMY replied:
The issue is that the method is not allowed to return null, not the annotation itself. use `return new Object();`.

NEZNAMY replied:
This isn't resolved.

NEZNAMY commented:
How does this work exactly? You write the packet the way it was seen at some point and via remaps it? That wouldn't explain the need for so many classes though.

Rubenicos commented:
ViaVersion use "the best" or "the correct way" to remap packets across versions, for example, if your server is on 1.16.5 and a player on 1.21.4 connects, any packet that is sent to player first need to be converted from 1.16.5 -> 1.17.x -> 1.18.x -> ... etc. until it reach to player version, that is the main reason it uses packages like `v1_19_1to1_19_3` for **every** protocol update.

If you create a packet using 1.17 protocol and send it to a 1.21.4 player, ViaVersion will do all the thing to make that packet reach the player.

So, I selected every protocol that makes a change on a packet and make a packet write for its changes, for example, the packet `PLAYER_INFO` is used to send tablist information and its changes are:
* 1.16 - RGB colors.
* 1.19.3 - Separated packet to remove entries, `BitSet` usage for actions and `UPDATE_LISTED` action.
* 1.21.2 - `UPDATE_LIST_ORDER` action.
* 1.21.4 - Shadow color and `UPDATE_HAT` action.

Next, I made an extension from abstract class (that contains the shared logic between version changes), this is because I see some classes on bukkit module named like "PacketTabList1193", but I can made it using nested classes or a protocol version check in the same class as well, tell me what do you think.

The other thing is the fabulous way how ViaVersion work, if ViaVersion gets updated, this changes will still working using the same TAB version, since ViaVersion "stacks" new protocols above old ones (following "the correct way") by adding a new protocol remapper without affecting the protocols that are currently used.

NEZNAMY commented:
So can we not just create a packet for the latest version and let Via remap it back? Or would that also require ViaBackwards which is why you didn't do it?

NEZNAMY commented on component/src/main/java/me/neznamy/chat/component/TabComponent.java:
```diff
@@ -102,6 +103,11 @@ public abstract class TabComponent {
+    /** ViaVersion component from this component */
+    @Nullable
+    @Setter
+    private Object viaComponent;
```
Why is this an object when it's always JsonObject? Same for all related methods.

Rubenicos replied:
Is an `Object` due is not a `com.google.gson.JsonObject`, it is a `com.viaversion.viaversion.libs.gson.JsonObject`

NEZNAMY replied:
And what's the problem?

Rubenicos replied:
If ViaVersion is not loaded it might fail with `NoClassDefFoundError`, I guess.

NEZNAMY replied:
It doesn't do that for adventure component though.

NEZNAMY commented on component/src/main/java/me/neznamy/chat/hook/ViaVersionHook.java:
```diff
@@ -0,0 +1,117 @@
+            return convert((KeybindComponent) component);
+        } else {
+            throw new IllegalStateException("Unknown component type: " + component.getClass().getName());
+        }
```
Why splitting the methods? Component style and extra are not limited to text component, just insert the key & value based on component type and always insert style.

NEZNAMY commented on component/src/main/java/me/neznamy/chat/hook/ViaVersionHook.java:
```diff
@@ -0,0 +1,117 @@
+        if (Boolean.TRUE.equals(modifier.getBold())) {
+            object.addProperty("bold", true);
+        }
+        object.addProperty("italic", Boolean.TRUE.equals(modifier.getItalic()));
```
Is this intentionally different than other magic codes?

Rubenicos replied:
It's a "bug" that exists since raw json text was created by Mojang, if you don't specify the text is not italic, it will show as italic text by default.

NEZNAMY replied:
Actually, there is a difference between false and null. Not specifying (null) instead of false would return in magic codes not being reset when they are supposed to be. Right?

Rubenicos commented:
> So can we not just create a packet for the latest version and let Via remap it back? Or would that also require ViaBackwards which is why you didn't do it?

Indeed, it is preferable to use only ViaVersion logic since this changes aims to achieve new features compatibility with players in newer versions.

NEZNAMY commented on component/src/main/java/me/neznamy/chat/hook/ViaVersionHook.java:
```diff
@@ -0,0 +1,117 @@
+        final ChatModifier modifier = component.getModifier();
+        if (modifier.getColor() != null) {
+            final TextColor color = modifier.getColor();
+            if (color.isLegacy()) {
```
I'm not 100% sure about this one. It might need a version check. Plugin might try to use a RGB code and then send it to 1.13 player this way.

Rubenicos replied:
I made a separated component send on ViaScoreboard13 to avoid RGB usage on pre 1.16 players.

NEZNAMY replied:
Then what's the point of this check? If legacy colors are used, they are converted to TextColor, which contains the correct rgb code that can be used.

Rubenicos replied:
It's a client-side micro-optimization to send (for example) `"color": "white"` instead of `"color": "#ffffff"`

NEZNAMY replied:
Realistically, is it? It just adds more code here, which can confuse someone.

NEZNAMY commented on viaversion/src/main/java/me/neznamy/tab/platforms/viaversion/platform/ViaVersionPlatform.java:
```diff
@@ -0,0 +1,95 @@
+
+    @NotNull
+    public static <P extends TabPlayer> FunctionWithException<P, Scoreboard> findScoreboardProvider(@NotNull ProtocolVersion serverVersion) {
+        FunctionWithException<P, Scoreboard> function = FunctionWithException.empty();
```
Can't you just return immediately without assigning to variable? (will require swapping the checks)

Rubenicos replied:
Do you mean something like this?
```java
if (serverVersion.getNetworkId() < ProtocolVersion.V1_13.getNetworkId()) {
    return player -> {
        if (player.getVersion().getNetworkId() >= ProtocolVersion.V1_20_3.getNetworkId()) {
            return new ViaScoreboard1203<>(player);
        } else if (player.getVersion().getNetworkId() >= ProtocolVersion.V1_16.getNetworkId()) {
            return new ViaScoreboard16<>(player);
        } else if (player.getVersion().getNetworkId() >= ProtocolVersion.V1_13.getNetworkId()) {
            return new ViaScoreboard13<>(player);
        }
        return null;
    };
} else if (serverVersion.getNetworkId() < ProtocolVersion.V1_16.getNetworkId()) {
    return player -> {
        if (player.getVersion().getNetworkId() >= ProtocolVersion.V1_16.getNetworkId()) {
            return new ViaScoreboard16<>(player);
        } else if (player.getVersion().getNetworkId() >= ProtocolVersion.V1_13.getNetworkId()) {
            return new ViaScoreboard13<>(player);
        }
        return null;
    };
} else if (serverVersion.getNetworkId() < ProtocolVersion.V1_20_3.getNetworkId()) {
    return player -> {
        if (player.getVersion().getNetworkId() >= ProtocolVersion.V1_20_3.getNetworkId()) {
            return new ViaScoreboard1203<>(player);
        }
        return null;
    };
}
return FunctionWithException.empty();
```

NEZNAMY replied:
Yes, something like that.

NEZNAMY commented on component/src/main/java/me/neznamy/chat/hook/ViaVersionHook.java:
```diff
@@ -0,0 +1,113 @@
+public class ViaVersionHook {
+
+    @NotNull
+    public static Object convert(@NotNull TabComponent component) {
```
You could merge all of these methods into one.

NEZNAMY commented on viaversion/src/main/java/me/neznamy/tab/platforms/viaversion/scoreboard/ViaScoreboard1203.java:
```diff
@@ -0,0 +1,103 @@
+    protected void writeObjectiveDisplay(@NonNull PacketWrapper packet, @NonNull Objective objective) {
+        super.writeObjectiveDisplay(packet, objective);
+        // ViaVersion configuration compatibility
+        if (objective.getHealthDisplay() == HealthDisplay.INTEGER && Via.getConfig().hideScoreboardNumbers()) {
```
We don't want to respect this viaversion option. Currently, it causes `score` to not be visible for 1.20.3+ players on <1.20.3 servers, but not only in sidebar, but also belowname and playerlist. This is an opportunity to resolve this problem.

Rubenicos commented:
In summary:
* I forgot to send components as NBT for +1.20.3 players, so I created another method on TabComponent (Still need to test if a NoClassDefFoundError occurs or not).
* Added a dummy object for unused scoreboard team.
* Added safe check to get skin data from bukkit player.

Still having some errors due TAB send tablist and scoreboard information too early (before ViaVersion do all the stuff to properly send packets to players) causing to not see any of both somethings, can be fixed for scoreboard if a delay is set, but with tablist I didn't see an delay on default configuration.

NEZNAMY commented on viaversion/src/main/java/me/neznamy/tab/platforms/viaversion/scoreboard/ViaScoreboard.java:
```diff
@@ -158,9 +158,11 @@ protected void sendTeamUpdate(int mode, @NonNull Team team) {
         packet.scheduleSend(protocol);
     }
 
+    protected abstract void writeComponent(@NonNull PacketWrapper packet, @NonNull String text);
```
Can replace usage with `new SimpleTextComponent(text)` and no need for 2 methods.

NEZNAMY commented on viaversion/src/main/java/me/neznamy/tab/platforms/viaversion/tablist/ViaTabList1214.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    protected void writeEntry(@NonNull PacketWrapper packet, @NonNull Entry entry) {
+        super.writeEntry(packet, entry);
+        packet.write(Types.VAR_INT, entry.getListOrder());
```
1212 already writes list order.

NEZNAMY commented on viaversion/src/main/java/me/neznamy/tab/platforms/viaversion/tablist/ViaTabList16.java:
```diff
@@ -0,0 +1,116 @@
+        }
+        packet.write(Types.VAR_INT, entry.getGameMode());
+        packet.write(Types.VAR_INT, entry.getLatency());
+        packet.write(Types.OPTIONAL_COMPONENT, entry.getDisplayName() == null ? null : entry.getDisplayName().toViaVersion());
```
can use #writeOptionalComponent

NEZNAMY commented:
If you move the classes into Bukkit module, you'll be able to use `SkinData` class to properly get player's skin.

NEZNAMY commented:
1.21.5-pre1 is out. We can expect the full release within the next weeks. Do you think we could get this PR fully resolved by then?

Rubenicos commented:
Should I move viaversion module into bukkit module maintaining the package `me.neznamy.tab.platforms.viaversion` or move to `me.neznamy.tab.platforms.bukkit`?

NEZNAMY commented:
Move it to bukkit package.

NEZNAMY commented:
Maybe moving the classes (scoreboard & tablist) to sub-packages called viaversion would be a good idea to make it not be in the same package as other implementations.

I have also noticed some inconsistency. You call 1.13 and 1.16 classes `13` and `16`, while calling for example 1.20.3 `1203`. Shouldn't they be called `113` and `116` for consistency?

NEZNAMY commented:
Yes.

NEZNAMY commented:
Were you able to make any progress regarding the issue on join you described? Perhaps viaversion people might be able to help?

Rubenicos commented:
Added a first-time 0.5 seconds delay using player channel event loop, after some tests seems to be needed 0.4 seconds or more to avoid any error, so I used 0.5 to make sure it works.

NEZNAMY commented:
That sounds very unreliable. Is there no way to detect this? Do none of the variables inside viaversion change?

NEZNAMY commented on bukkit/src/main/java/me/neznamy/tab/platforms/bukkit/tablist/viaversion/ViaTabList1193.java:
```diff
@@ -0,0 +1,105 @@
+        // Players
+        packet.write(Types.UUID_ARRAY, new UUID[] { entry });
+
+        packet.scheduleSend(protocol);
```
Forgot to call delayed send here?

Rubenicos commented:
> That sounds very unreliable. Is there no way to detect this? Do none of the variables inside viaversion change?

I checked the pipeline, and everything seems to be fine (ViaVersion is already registered when ViaTabList and ViaScoreboard constructor is called), so I think this is a netty-related behaviour.

> Also this may result in packets being sent out of order when sent just after the delay.

* The player joins
* TAB send multiple packets in the same thread
* ViaTabList / ViaScoreboard make a delayed execution for those packets (since they are in the same thread) using `PacketWrapper#send()` to send them immediately after delay
* After delay any packet is sent using `PacketWrapper#scheduleSend()` in the same event loop that is used for delayed executions, so technically they will not interfere with packets that was sent before

NEZNAMY commented:
So let's say at 400ms you send a packet. There will be a 500ms delay, resulting in it being sent at 900ms. What if a packet is sent at 600ms? Will it be sent after or before?

Rubenicos commented:
Using a `Queue` is more consistent, if I could find a way to detect that when the channel context will correctly send the packet I could make a better solution

NEZNAMY commented:
Everything looks fine to me now. Have all server-client version pairs been tested?

Rubenicos commented:
Tested by:
* Joining on 1.8 server using 1.21.4, 1.21.2, 1.19.4 and 1.16.5
* Joining on 1.12 server using 1.21.4, 1.21.2, 1.19.4 and 1.16.5
* Joining on 1.17.1 server using 1.21.4, 1.21.2 and 1.19.4
* Joining on 1.20.1 server using 1.21.4, and 1.21.2

Everything using default TAB configuration + enabling scoreboard

NEZNAMY commented:
Have you tried 1.13 client too? For scoreboard. Otherwise, sounds good.

Rubenicos commented:
Now tested by:
* Joining on 1.8 server using 1.13.2
* Joining on 1.12.2 server using 1.13.2

NEZNAMY commented:
Anything else before merging?

Rubenicos commented:
Anything else.

Btw, I known unlimited nametag mode was removed since v5.
My server currently use an additional plugin for that, but I think this can be achieved in a cleaner way by providing an empty API inside TAB code.
For example, if TAB doens't found a UnlimitedNametag provider (that will never find inside the TAB code) it will work as normal.

NEZNAMY commented:
What's the point of adding an empty interface that will never be instantiated?

Rubenicos commented:
To allow other people to register its own implementation, for example, an implementation compatible with ModelEngine that shows models as name decorations

NEZNAMY commented:
I think that is quite of out of scope, especially for a feature that was literally removed. I have no interest in doing anything related to it.

Rubenicos commented:
Understandable

NEZNAMY commented:
Thank you for your contribution.

NEZNAMY commented:
I have seen someone report the delay sometimes not being enough. After printing objects of the connection before and after the delay, I noticed one difference:

```
protocolInfo: ProtocolInfo{clientState=LOGIN, serverState=PLAY, protocolVersion=1.21.4 (769), serverProtocolVersion=1.8.x (47), username='_NEZNAMY_', uuid=6603bf58-226d-3b2b-8f5e-2f5d89104533}

protocolInfo: ProtocolInfo{clientState=PLAY, serverState=PLAY, protocolVersion=1.21.4 (769), serverProtocolVersion=1.8.x (47), username='_NEZNAMY_', uuid=6603bf58-226d-3b2b-8f5e-2f5d89104533}
```

Perhaps the reason why it didn't work is because those packets cannot be sent during login/configuration state, which for some reason lasts way longer than it should. Would you have any idea how to solve this with some kind of event listening?

Rubenicos commented:
Brilliant!
I now know how to send the packets properly

I'll make a pull request in a bit

NEZNAMY commented on bukkit/src/main/java/me/neznamy/tab/platforms/bukkit/tablist/viaversion/ViaTabList116.java:
```diff
@@ -0,0 +1,114 @@
+            // Properties size
+            packet.write(Types.VAR_INT, 1);
+            // Properties
+            packet.write(Types.STRING, entry.getName());
```
This was supposed to be "textures", not player name.

NEZNAMY commented on bukkit/src/main/java/me/neznamy/tab/platforms/bukkit/tablist/viaversion/ViaTabList1193.java:
```diff
@@ -0,0 +1,105 @@
+            // Properties size
+            packet.write(Types.VAR_INT, 1);
+            // Properties
+            packet.write(Types.STRING, entry.getName());
```
Same here.

Rubenicos commented:
I understand, seeing the situation this will only bring future problems, I implemented ViaVersion based on its behaviour on packet sending, but it seems that this differs from the game's own protocol in certain aspects, so a complete investigation of Minecraft client code would be required to make something without errors.

By removing this feature you could leave a getter and setter for `serverImplementationProvider` field in the `BukkitPlatform`?
I'd at least like to leave ViaVersion addon as a personal project.

NEZNAMY commented:
That would not solve your request, because there are things you must take from the server itself, such as player skin. But I guess you will find out. Either way, I can do that one exact thing for you if you want.
